powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/callbacks/CallbackUrlEventService.java

```diff
@@ -0,0 +1,287 @@
+    }
+
+    /**
+     * Dispatch Callback URL Event.
+     * @param callbackUrlEventEntity Event to dispatch.
+     */
+    private void dispatchPendingCallbackUrlEvent(final CallbackUrlEventEntity callbackUrlEventEntity) {
+        callbackUrlEventEntity.setStatus(CallbackUrlEventStatus.PROCESSING);
+        callbackUrlEventEntity.setTimestampNextCall(null);
+        callbackUrlEventEntity.setTimestampLastCall(LocalDateTime.now());
+        final CallbackUrlEventEntity savedEventEntity = callbackUrlEventRepository.save(callbackUrlEventEntity);
+
+        final CallbackUrlEvent callbackUrlEvent = CallbackUrlConvertor.convert(savedEventEntity);
+        TransactionUtils.executeAfterTransactionCommits(
+                () -> postCallback(callbackUrlEvent)
+        );
+    }
+
+    /**
+     * Send Callback URL Event as a non-blocking POST request.
+     * @param callbackUrlEvent Event to post.
+     */
+    private void postCallback(final CallbackUrlEvent callbackUrlEvent) {
+        if (callbackUrlEvent.status() != CallbackUrlEventStatus.PROCESSING) {
+            logger.debug("Callback URL Event is not in PROCESSING state: callbackUrlEventId={}", callbackUrlEvent.callbackUrlEventEntityId());
+            return;
+        }
+
+        try {
+            final Consumer<ResponseEntity<String>> onSuccess = response -> callbackUrlEventResponseHandler.handleSuccess(callbackUrlEvent);
+            final Consumer<Throwable> onError = error -> callbackUrlEventResponseHandler.handleFailure(callbackUrlEvent, error);
+            final ParameterizedTypeReference<String> responseType = new ParameterizedTypeReference<>(){};
+
+            final RestClient restClient = getRestClient(callbackUrlEvent.callbackUrlEventConfiguration());
+            final MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
+            headers.add("Idempotency-Key", callbackUrlEvent.idempotencyKey());
+
+            restClient.postNonBlocking(callbackUrlEvent.callbackUrlEventConfiguration().callbackUrl(),
+                    callbackUrlEvent.callbackData(),
+                    new LinkedMultiValueMap<>(),
+                    headers,
+                    responseType,
+                    onSuccess,
+                    onError);
+
+            logger.debug("CallbackUrlEvent {} was dispatched.", callbackUrlEvent.callbackUrlEventEntityId());
+
+        } catch (RestClientException e) {
+            callbackUrlEventResponseHandler.handleFailure(callbackUrlEvent, e);
+        }
+    }
+
+    /**
+     * Get default force rerun period, after which is a Callback URL Event in PROCESSING state considered stale.
+     * @return Default force rerun period.
+     */
+    private Duration defaultForceRerunPeriod() {
+        // This is an arbitrary value, representing allowed delay before trying establishing remote connection.
+        final Duration allowedProcessingDelay = Duration.ofSeconds(10);
```

Comment:
We should use max (arbitrary value, backoff period). We need to be sure the legitimate time of the request processing passed, but we also should not process the request sooner than his backoff period in case of processed failure.

scenario:
- we execute the callback
- somebody reset the server before the standard answer
- reset stale job will update all events that have last call timestamp older than now - forcedRerun.
  - if the call was successful we do not mind to wait longer period - as the receiver was already informed.
  - if the call was not successful we should wait the same period i.e. add backoff interval.

Reply:
I consider server restart / crash as a rare event, which should affect the callback processing as little as possible and  not consider it as a callback processing failure.
Do you suggest to behave to stale events in the same way as to failed events? i.e. incrementing attempts and set backoff?